lib/control/control-server.c

```diff
@@ -41,11 +166,27 @@ void
 control_server_init_instance(ControlServer *self, const gchar *path)
 {
   self->control_socket_name = g_strdup(path);
+  self->worker_threads = NULL;
+}
+
+static void
+_delete_thread_command_runner(gpointer data)
+{
+  ThreadedCommandRunner *self = (ThreadedCommandRunner *) data;
+  g_assert(self->real_thread.tid_saved == TRUE);
+  thread_cancel(self->real_thread.tid);
+  g_thread_join(self->thread);
+  _thread_command_runner_free(self);
 }
 
 void
 control_server_free(ControlServer *self)
 {
+  // it's not racy as the iv_main() is executed by this thread
+  // posted events are not executed as iv_quit() is already called (before control_server_free)
+  // but it is possible that some ThreadedCommandRunner::thread are still running
+  if (self->worker_threads)
+    g_list_free_full(self->worker_threads, _delete_thread_command_runner);
```

Comment:
Unfortunately this does not cancel the long running commands. `control_server_free` is called too late: after main loop is not running. For example, syslog-ng-ctl stats is running in mainloop, so main_loop_deinit will not be called until it is finished.

You can try this:

```
@version: 3.29

options {
 stats-level(3);
};

log {
  source { example-msg-generator(freq(0) num(100000000)
                                 values("HOST" => "host$(iterate $(+ 1 $_) 0)") keep-hostname(yes)); };
  destination { file(/dev/null); };
};
```

Let syslog-ng to run a bit: example-msg-generator will generate a bit slowly, the longer you wait, the longer stats runs. After waiting a bit, call syslog-ng-ctl stats and syslog-ng-ctl stop at the same time. syslog-ng-ctl-stop will "hang".

Reply:
@furiel : Thanks!!
(It works perfectly with my original branch, when control server is run in a different thread - I'll fix it soon).

Reply:
Unfortunately only my original solution works... the reason is very simple: `stats_lock`...
When generator source (or any other source executed by the main thread) produces messages, it also updates its stat counters.  When the stats mutex is locked by a `syslog-ng-ctl stats` then ivykis will wait the actual 'job' to be finished (which can happen only after the mutex is unlocked) - until that it won't process the events of the control loop...
This is why we need to execute `ControlServer` in its own thread: that won't lock the stats mutex.